### PR TITLE
Add Azure SQL Database Failover Group REST API specs

### PR DESCRIPTION
**Pull request by btasdoven**
Adding Azure SQL Database Failover Group REST API specifications as the feature will be in public preview soon.

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [ ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ ] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 1]
```diff
@@ -0,0 +1,616 @@
...
+                        "schema": {
+                            "$ref": "#/definitions/ArmErrorResponseMessage"
```
Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Remove this

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+                        }
+                    }
+                },
+                "deprecated": false
+            },
+            "put": {
+                "tags": [
+                    "FailoverGroups"
+                ],
+                "description": "Creates or updates the failover group with the given name.",
```
Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Creates or updates a failover group .

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+                    }
+                },
+                "deprecated": false,
+                "x-ms-long-running-operation": true
+            },
+            "delete": {
+                "tags": [
+                    "FailoverGroups"
+                ],
+                "description": "Deletes the failover group with the given name.",
```
Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Deletes a failover group.

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 4]
```diff
@@ -0,0 +1,616 @@
...
+                        "schema": {
+                            "$ref": "#/definitions/ArmErrorResponseMessage"
```
Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Remove

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+    "produces": [
+        "application/json"
+    ],
+    "paths": {
+        "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/failoverGroups/{failoverGroupName}": {
+            "get": {
+                "tags": [
+                    "FailoverGroups"
+                ],
+                "description": "Gets the failover group with the given name.",
```
Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Gets a failover group.

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+                        "schema": {
+                            "$ref": "#/definitions/ArmErrorResponseMessage"
```
Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Remove

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+                    }
+                },
+                "deprecated": false,
+                "x-ms-long-running-operation": true
+            },
+            "patch": {
+                "tags": [
+                    "FailoverGroups"
+                ],
+                "description": "Partially updates the failover group with the given name.",
```
Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Updates a failover group.

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 8]
```diff
@@ -0,0 +1,616 @@
...
+                        "schema": {
+                            "$ref": "#/definitions/ArmErrorResponseMessage"
```
Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Remove

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 9]
```diff
@@ -0,0 +1,616 @@
...
+                "x-ms-long-running-operation": true
+            }
+        },
+        "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/failoverGroups": {
+            "get": {
+                "tags": [
+                    "FailoverGroups"
+                ],
+                "description": "Lists the failover groups on the given server.",
+                "operationId": "FailoverGroups_List",
```
Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: FailoverGroups_ListByServer

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 10]
```diff
@@ -0,0 +1,616 @@
...
+                "deprecated": false,
+                "x-ms-long-running-operation": true
+            }
+        },
+        "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/failoverGroups": {
+            "get": {
+                "tags": [
+                    "FailoverGroups"
+                ],
+                "description": "Lists the failover groups on the given server.",
```
Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Lists the failover groups in a server.

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 11]
```diff
@@ -0,0 +1,616 @@
...
+                },
+                "deprecated": false
+            }
+        },
+        "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Sql/servers/{serverName}/failoverGroups/{failoverGroupName}/failover": {
+            "post": {
+                "tags": [
+                    "FailoverGroups"
+                ],
+                "description": "Starts planned failover operation for the failover group.",
```
Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: I assume that like replicationLinks/failover this is called on the new primary, is this correct? If so, please describe. Below is what I wrote in Azure CLI for this operation, consider something similar. 

Command
    az sql db replica set-primary: Sets which replica database is primary by failing over from the
    current primary replica database.

The swagger doc for replicationLinks/failover probably needs to be improved as well.

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json (hunk 8 above)
Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: This should be removed everywhere, I will stop commenting on the later ones :)

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 12]
```diff
@@ -0,0 +1,616 @@
...
...r.\n\n * 400 InvalidFailoverGroupRegion - Servers specified in a Failover Group need to reside in different regions to provide isolation.\n\n * 400 ServerNotFound - The requested server was not found.\n\n * 400 FailoverGroupDoesNotExist - Failover group does not exist on a server.\n\n * 400 FailoverGroupNotSecondary - Failover cannot be initiated from the primary server in a Failover Group.\n\n * 400 PlannedFailoverTimedOutForDatabase - User invoked planned failover, it timed out, and a specific database appears to be to blame.\n\n * 400 PlannedFailoverTimedOut - User invoked planned failover, and it timed out while trying to contact the partner management service.\n\n * 400 DatabaseInFailoverGroupNotPrimary - GeoDR link for the database already exists, but its role is not 'Primary' or it is not continuous copy link.\n\n * 400 InvalidAddSecondaryPermission - User does not have sufficient permission to add secondary on the specified server.\n\n * 400 InvalidSku - The user specified an invalid sku.\n\n * 400 FeatureDisabledOnSelectedEdition - User attempted to use a feature which is disabled on current database edition.\n\n * 400 InvalidTargetSubregion - The target server of a non-re
```
Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Remove all these

Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: please remove all these and use "allOf" to include properties from ProxyResource or TrackedResource. You can see examples of this in other sql swagger files. You will need to copy Resource and ProxyResource/TrackedResource definitions into this file.

Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: This should just be called 'FailoverGroupProperties'. 'Resource' is redundant.

Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: What are possible values? Describe with enum. Is it readonly?

Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Readonly?

Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Readonly?

Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Needs x-ms-enum. See https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/creating-swagger.md#Enum-x-ms-enum

Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Need x-ms-enum

Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: FailoverGroupList

**Review by jaredmoo — CHANGES_REQUESTED** [review]

**Comment by dsgouda** [comment]
I shall take a look at this soon

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 13]
```diff
@@ -0,0 +1,600 @@
...
+  "schemes": [
+    "https"
+  ],
+  "consumes": [
+    "application/json"
+  ],
+  "produces": [
+    "application/json"
+  ],
+  "paths": {
```
Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Please expose an ```/operations``` API for this spec

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 14]
```diff
@@ -0,0 +1,600 @@
...
+        "description": "Gets a failover group.",
+        "operationId": "FailoverGroups_Get",
+        "x-ms-examples": {
+          "Get failover group": {
+            "$ref": "../examples/FailoverGroupGet.json"
+          }
+        },
+        "parameters": [
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
```
Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: nit: For consistency, put ```SubscriptionId``` and ```ApiVersion``` together at the beginning or at the end of the parameters section

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 15]
```diff
@@ -0,0 +1,600 @@
...
+        "description": "Lists the failover groups in a server.",
+        "operationId": "FailoverGroups_ListByServer",
+        "x-ms-examples": {
+          "List failover group": {
+            "$ref": "../examples/FailoverGroupList.json"
+          }
+        },
+        "parameters": [
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
```
Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: nit: parameter ordering

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 16]
```diff
@@ -0,0 +1,600 @@
...
+        "description": "Sets which server is primary by failing over from the current primary server. This operation might result in data loss.",
+        "operationId": "FailoverGroups_ForceFailoverAllowDataLoss",
+        "x-ms-examples": {
+          "Forced failover failover group": {
+            "$ref": "../examples/FailoverGroupForceFailoverAllowDataLoss.json"
+          }
+        },
+        "parameters": [
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
```
Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: nit: parameter ordering

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json [hunk 17]
```diff
@@ -0,0 +1,600 @@
...
...n\n * 409 FailoverGroupDnsRecordInUse - A duplicate DNS record exists for the requested endpoint.\n\n * 409 InvalidFailoverGroupName - Invalid Failover Group name was supplied.\n\n * 409 SubscriptionDisabled - Subscription is disabled.\n\n * 409 ConflictingSystemOperationInProgress - A system maintenance operation is in progress on the database and further operations need to wait until it is completed.\n\n * 409 InvalidOperationForDatabaseNotInReplicationRelationship - A replication seeding operation was performed on a database that is already in a replication relationship.\n\n * 409 InvalidDatabaseStateForOperation - The operation is not allowed on the database in its current replication state.\n\n * 409 DuplicateGeoDrRelation - The databases are already in a replication relation. This is a duplicate request.\n\n * 409 GeoReplicationCannotBecomePrimaryDuringUndo - User attempted to failover or force-terminate a geo-link while the secondary is in a state where it may not be physically consistent and so cannot enter the primary role.\n\n * 429 SubscriptionTooManyRequests - Requests beyond max requests that can be processed by available resources.\n\n * 429 ConflictingServerOperation
```
Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Consider marking this as ```"modelAsString": true``` if the enum is expected to change in the future, not doing so causes breaking changes

Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Same comment as above for enums

Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: This also seems like something that the server is setting and should not be a part of the request. Could this be marked as readonly

Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Same comment as above for enums

Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Some properties here look like candidates for readonly properties, please verify if you have all properties correctly marked as readonly

Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: nit: If you're not using this, you can remove it

**Review by dsgouda — CHANGES_REQUESTED** [review]

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json (hunk 17 above)
Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Consider using ```x-ms-client-flatten``` here for the nested properties for better user experience

**Review by dsgouda — CHANGES_REQUESTED** [review]

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json (hunk 13 above)
Comment (jaredmoo) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Hey @dsgouda , I was told by @veronicagg that /operations isn't needed in each swagger file - it's just needed once per RP. https://github.com/Azure/azure-rest-api-specs/pull/1005#issuecomment-285826237

Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: That's correct, if there is an operations API already existing which returns the operations mentioned in this spec, you can disregard this comment

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json (hunk 17 above)
Comment (btasdoven) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: I can't correctly see which lines you comment on @dsgouda . This and all of the next comments are for the line 321. Any ideas to view them?

Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Weird, I might have to try this manually, let me get back in a while

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json (hunk 14 above)
Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Also applies [here](https://github.com/Azure/azure-rest-api-specs/pull/1079/files#diff-090eb226c435d8d8ba061faca6869009R72), [here](https://github.com/Azure/azure-rest-api-specs/pull/1079/files#diff-090eb226c435d8d8ba061faca6869009R72), [here](https://github.com/Azure/azure-rest-api-specs/pull/1079/files#diff-090eb226c435d8d8ba061faca6869009R123), [here](https://github.com/Azure/azure-rest-api-specs/pull/1079/files#diff-090eb226c435d8d8ba061faca6869009R165), [here](https://github.com/Azure/azure-rest-api-specs/pull/1079/files#diff-090eb226c435d8d8ba061faca6869009R212), [here](https://github.com/Azure/azure-rest-api-specs/pull/1079/files#diff-090eb226c435d8d8ba061faca6869009R252) and [here](https://github.com/Azure/azure-rest-api-specs/pull/1079/files#diff-090eb226c435d8d8ba061faca6869009R296)

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json (hunk 17 above)
Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Linked [here ](https://github.com/Azure/azure-rest-api-specs/pull/1079/files#diff-090eb226c435d8d8ba061faca6869009R351)

Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Linked [here](https://github.com/Azure/azure-rest-api-specs/pull/1079/files#diff-090eb226c435d8d8ba061faca6869009R471) and [here](https://github.com/Azure/azure-rest-api-specs/pull/1079/files#diff-090eb226c435d8d8ba061faca6869009R492)

Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Linked [here](https://github.com/Azure/azure-rest-api-specs/pull/1079/files#diff-090eb226c435d8d8ba061faca6869009R476), [here](https://github.com/Azure/azure-rest-api-specs/pull/1079/files#diff-090eb226c435d8d8ba061faca6869009R523) and [here](https://github.com/Azure/azure-rest-api-specs/pull/1079/files#diff-090eb226c435d8d8ba061faca6869009R497)

Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Linked [here](https://github.com/Azure/azure-rest-api-specs/pull/1079/files#diff-090eb226c435d8d8ba061faca6869009R416)

**Comment by dsgouda** [comment]
Github is probably acting up because of the super long description for one of the operation responses, I've tried my best to link my comments to the line numbers, hope that helps

File: arm-sql/2015-05-01-preview/swagger/failoverGroups.json (hunk 17 above)
Comment (btasdoven) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: We allow customers to update those fields. I made a second pass on them to make sure all the read-only properties are marked as read only.

Comment (btasdoven) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Thanks!

Comment (dsgouda) [inline arm-sql/2015-05-01-preview/swagger/failoverGroups.json]: Just curious, is FailoverGroupReadWriteEndpoint some sort of a url?

**Review by dsgouda — APPROVED** [review]

**Comment by dsgouda** [comment]
Linter rules look good, generated C# code seems to compile too. LGTM

**Pull request merged**